helix-view/src/document.rs

```diff
@@ -859,24 +929,46 @@ impl Document {
                     .unwrap();
             }
 
+            let changes = transaction.changes();
+
             // map state.diagnostics over changes::map_pos too
             for diagnostic in &mut self.diagnostics {
-                use helix_core::Assoc;
-                let changes = transaction.changes();
                 diagnostic.range.start = changes.map_pos(diagnostic.range.start, Assoc::After);
                 diagnostic.range.end = changes.map_pos(diagnostic.range.end, Assoc::After);
                 diagnostic.line = self.text.char_to_line(diagnostic.range.start);
             }
             self.diagnostics
                 .sort_unstable_by_key(|diagnostic| diagnostic.range);
 
+            // Update the inlay hint annotations' positions, helping ensure they are displayed in the proper place
+            let apply_inlay_hint_changes = |annotations: &mut Rc<[InlineAnnotation]>| {
+                if let Some(data) = Rc::get_mut(annotations) {
+                    for inline in data.iter_mut() {
+                        inline.char_idx = changes.map_pos(inline.char_idx, Assoc::After);
```

Comment:
Neither `Assoc::After` or `Assoc::Before` works well here that is because both of these were designed for **ranges** where the start is mapped with `Assoc::Before` and `Assoc::After`. However for inlay hints we want something different. Inlay hints are bound to a single point so when that point is removed when also want to delete the inlay hint. For that you need to add a new variant `Assoc::Point` which returns `None` in case the position which contains `char_idx` was removed/replaced.

For that to worth tough we need a way to remove `InlineAnnotations` so the equivalent for `Vec::retain` but for `Rc<[T]>` which is not possible without a custom reference counted type (which we probably don't want to add here).

I am considering just switching `TextAnnotations` back to using slices instead of `Rc`  which it originally did. I changed it to make the API easier to use/remove lifetime issues that occured in #5340 but I think those issues were actually resolved by something else so maybe a slice is a better choice afterall.

I don't have more time to look into this right now but you could start with implementing `Assoc::Delete` and I will try to comeup with something in the next couple of days.

Reply:
I'm not sure I want to do this in this PR. Nor that I am qualified to do it.

`ChangeSet::map_pos()` currently unconditionnally returns a `usize`, adding `Assoc::Delete` or `Assoc::Point` would need to change the return type to account for the deletion cleanly.

Instead I think this PR should either be blocked until `map_pos` and `Assoc::...` are better equipped or, if we don't want to block too long, make an issue for it and fix it later on since for now `Assoc::After` works well enough (it probably has strange edge cases at the end of files but apart from that it should be okay for a v1)

Reply:
If you don't feel confident working on this that's fine. I will create a PR myself in the next couple days that that you can then depend on. It's not that big of a change so I dlunt it will block this PR for long.

I think it is a requirement for this because without properly handeling this some odd effects do occur. For example if you delete two lines `2xd` then all inlay hints within those lines are moved to the start of the next line

Reply:
I actually changed my mind on this. Mapping points doesn't work well in general. The right fix here would be to convert these points to ranges by expanding to the word at the specific position and then tracking that range (only removing it when empty just like we do for diagnostics).

I have a pretty good idea how to implement this but it's a bit more effort and I don't have the necessary time right now. In practice this not as large of a problem as I orginally tough because the type hints get refreshed relatively quickly (on idle timeout) so the artificats usually don't stick around for that long. So I think I can do the necessary refactors in a followup PR.